pkg/logs/internal/tag/provider.go

```diff
@@ -23,27 +22,31 @@ type Provider interface {
 	GetTags() []string
 }
 
+type tagFunction func(entity string, cardinality types.TagCardinality) ([]string, error)
```

Comment:
I take it that you're trying to avoid using the full `tagger.Component` interface here because it has lots of methods, but we only need `Tag`. Even still, I think it would be better to define a one off interface instead of using a function as a type:

```
type EntityTagAdder interface {
  Tag(...)
}
```

In `provider_benchmark_test.go` then `dummyTagFunction` should instead be dummyTagAdder which could implement this interface. In `provider_test.go` the `fakeTagger` will already fulfills this interface as is.

It looks like `pkg/logs/tailers/file/tailer.go` is already importing `pkg/logs/internal/tag` so maybe this `EntityTagAdder` interface should live there, and all of the changes in this PR could use that instead.

Reply:
That looks much cleaner and expandable if we aver want to support more tag functions in the future 😄